cmd/juju/machine/upgradeseries.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package machine
+
+import (
+	"strings"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/os/series"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/cmd/modelcmd"
+)
+
+// Actions
+const (
+	PrepareCommand  = "prepare"
+	CompleteCommand = "complete"
+)
+
+// NewUpgradeSeriesCommand returns a command which upgrades the series of
+// an application or machine.
+func NewUpgradeSeriesCommand() cmd.Command {
+	return modelcmd.Wrap(&upgradeSeriesCommand{})
+}
+
+type upgradeMachineSeriesAPI interface {
+}
+
+// upgradeSeriesCommand is responsible for updating the series of an application or machine.
+type upgradeSeriesCommand struct {
+	modelcmd.ModelCommandBase
+	// modelcmd.IAASOnlyCommand
+
+	// upgradeMachineSeriesClient upgradeMachineSeriesAPI
+
+	prepCommand   string
+	force         bool
+	machineNumber string
+	series        string
+}
+
+var upgradeSeriesDoc = `
+Upgrade a machine's operating system series.
+
+When a machine's series, the --force option should be used
+with caution since using a charm on a machine running an unsupported series may
+cause unexpected behavior. Alternately, if the requested series supported in
+later revisions of the charm, upgrade-charm can run beforehand.
+
+Examples:
+	juju upgrade-series prepare <machine> <series>
+	juju upgrade-series complete <machine>
+
+See also:
+    machines
+    status
+    upgrade-charm
+    set-series
+`
+
+func (c *upgradeSeriesCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "upgrade-series",
+		Args:    "<action> [args]",
+		Purpose: "Upgrade a machine's series.",
+		Doc:     upgradeSeriesDoc,
+	}
+}
+
+func (c *upgradeSeriesCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+	f.BoolVar(&c.force, "force", false, "Upgrade even if the series is not supported by the charm and/or related subordinate charms.")
+	f.BoolVar(&c.force, "agree", false, "Agree to upgrade the series. Otherwise, a confirmation prompt will be displayed after the command is run.")
+}
+
+// Init implements cmd.Command.
+func (c *upgradeSeriesCommand) Init(args []string) error {
+	numArguments := 3
+
+	if len(args) < 1 {
+		return errors.Errorf("wrong number of arguments")
+	}
+
+	prepCommandStrings := []string{PrepareCommand, CompleteCommand}
+	prepCommand, err := checkPrepCommands(prepCommandStrings, args[0])
+	if err != nil {
+		return errors.Annotate(err, "invalid argument")
+	}
+	c.prepCommand = prepCommand
+
+	if c.prepCommand == CompleteCommand {
+		numArguments = 2
+	}
+
+	if len(args) != numArguments {
+		return errors.Errorf("wrong number of arguments")
```

Comment:
How should we make it more specific? Perhaps, for example, " the completed command takes 2 arguments" or "Machine must be specified". I would feel more comfortable waiting until this story fleshes out a bit before fine tuning the user UX.

Reply:
I'd suggest that when we check if this is the complete command we in there validate that if the args < 2 we note "machine number to complete upgrade is missing" since we know that in detail there. 

Here we're less sure. They could have left out the prepare or complete command argument, the machine number target, or the series (if prepare). Anything we can do to help guide the user to the right missing bit is appreciate. 

If we want to move forward here the best thing is to note a bug on clarifying missing args and tag it with upgradeSeries so that we can work them down during feature development.

Reply:
https://bugs.launchpad.net/juju/+bug/1776903